### PR TITLE
Add tests for ControllerManager with non-default CM update rates

### LINKED ISSUE #541: [ControllerManager] Test improvement: tests different individual control rates and run them for a few iterations.
We should add tests with different rates and running them few iterations.

_Originally posted by @destogl in https://github.com/ros-controls/ros2_control/pull/513#r715550357_

### PR DESCRIPTION
**Pull request by AdvityaCode**
Closes #541

This addresses an older request from #513 to add tests that exercise the ControllerManager update loop at different rates across multiple iterations. I noticed this issue had been open since 2021 with no activity or open PR, and this is very closely related to a project I would like to work on in the future so I decided to share my idea for implementation. So if this is not what was expected, I am happy to learn and revise based on feedback!

**What was missing:**
All existing rate-related tests (`TestControllerUpdateRates`, `TestControllerManagerWithUpdateRates`) create the ControllerManager at the default 100 Hz and only vary the individual controller rate. No test exercised a CM running at a non-default rate (e.g. 200 Hz, 500 Hz, 1000 Hz).

**What this adds:**
A new parameterized test class`TestControllerManagerWithDifferentCmUpdateRates` that:
- Creates the CM at a specified non-default rate by passing `update_rate` as a CM node parameter
- Activates a controller at a sub-rate of the CM
- Runs the update loop for exactly `cm_update_rate` iterations (one simulated second) using the correct period for that rate
- Asserts the controller updated exactly `ctrl_update_rate` times

Test cases cover: 200/50 Hz, 200/100 Hz, 500/100 Hz, 500/50 Hz, 1000/100 Hz, 1000/500 Hz. All pairs are exact divisors to keep the expected count unambiguous.

### CODE REVIEW

**Review by bmagyar — DISMISSED** [review]
I'm a big fan or parametrized tests, thank you!

**Review by christophfroehlich — CHANGES_REQUESTED** [review]
Please have a look at the failing tests, and install pre-commit in your workspace

**Review by 4-legends — COMMENTED** [review]
@christophfroehlich @bmagyar  Will it make sense to add failure test cases such as controller manager rate is slower than controller update rate?

**Comment by christophfroehlich** [comment]
> @christophfroehlich @bmagyar Will it make sense to add failure test cases such as controller manager rate is slower than controller update rate?

If there is no such test yet (which I would have assumed), yes please.

**Comment by saikishor** [comment]
> @christophfroehlich @bmagyar Will it make sense to add failure test cases such as controller manager rate is slower than controller update rate?

I don't think we need this. This is already kinda tested here : https://github.com/ros-controls/ros2_control/blob/master/controller_interface/test/test_controller_interface.cpp#L101

**Comment by AdvityaCode** [comment]
Hi @christophfroehlich and @bmagyar,

 I've addressed the review feedback:

**Pre-commit:** Ran `pre-commit run --all-files` and committed the `clang-format` 
fixes.

**Failing tests:** The root cause was that the update loop was calling `cm_->update()` 
in a tight loop with a manually incremented `time_`, never sleeping. The CM's 
internal rate-divider uses the real trigger clock to decide when to fire sub-rate 
controllers, so without actual time passing the controllers never triggered. So,
`actual_updates` near 0 across all cases.

Fixed by following the same pattern as `TestControllerUpdateRates::check_the_controller_update_rate` 
— using `cm_->get_trigger_clock()->sleep_until()` between iterations and reading 
`get_trigger_clock()->now()` for the time argument. Also changed `EXPECT_EQ` to 
`EXPECT_NEAR(..., 1u)` to match the ±1 tolerance used in the existing rate tests 
for the first-cycle alignment edge case.

Also rebased onto upstream master. Let me know what you think.

File: controller_manager/test/test_controller_manager.cpp [hunk 1]
```diff
@@ -3624,3 +3624,105 @@ TEST_F(TestControllerManagerNotHandlingExceptions, controller_configure_on_excep
...
+  {
+    cm_->get_trigger_clock()->sleep_until(time_ + cm_duration);
+    time_ = cm_->get_trigger_clock()->now();
+    EXPECT_EQ(controller_interface::return_type::OK, cm_->update(time_, cm_duration));
+  }
+
+  const auto actual_updates = test_controller->internal_counter - initial_counter;
+  EXPECT_NEAR(actual_updates, ctrl_update_rate, 1u)
+    << "After " << cm_update_rate << " CM cycles at " << cm_update_rate << " Hz, a controller at "
+    << ctrl_update_rate << " Hz should have updated exactly " << ctrl_update_rate << " times";
```
[the remark this answers is not in the document]
Comment (christophfroehlich) [reply controller_manager/test/test_controller_manager.cpp:3720]: @AdvityaCode w.r.t. to your comment above, we should update the message here.

File: controller_manager/test/test_controller_manager.cpp [hunk 2]
```diff
@@ -3624,3 +3624,105 @@ TEST_F(TestControllerManagerNotHandlingExceptions, controller_configure_on_excep
...
+  auto switch_future = std::async(
+    std::launch::async, &controller_manager::ControllerManager::switch_controller, cm_,
+    start_controllers, stop_controllers, STRICT, true, rclcpp::Duration(0, 0));
+
+  ASSERT_EQ(std::future_status::timeout, switch_future.wait_for(std::chrono::milliseconds(100)))
+    << "switch_controller should be blocking until next update cycle";
+
+  // Use the real trigger clock so the CM's internal rate-division logic fires correctly
+  time_ = cm_->get_trigger_clock()->now();
+  cm_->get_trigger_clock()->sleep_until(time_ + PERIOD);
```
[the remark this answers is not in the document]
Comment (christophfroehlich) [reply controller_manager/test/test_controller_manager.cpp:3690]: @AdvityaCode do you agree?

**Review by christophfroehlich — CHANGES_REQUESTED** [review]
Tests are failing:
```

    [  FAILED  ] per_controller_update_rate_with_different_cm_rates/TestControllerManagerWithDifferentCmUpdateRates.per_controller_update_rate_with_different_cm_rates/4, where GetParam() = 8-byte object <E8-03 00-00 64-00 00-00>
    
     1 FAILED TEST
    -- run_test.py: return code 1
    -- run_test.py: inject classname prefix into gtest result file '/home/runner/work/ros2_control/ros2_control/.work/target_ws/build/controller_manager/test_results/controller_manager/test_controller_manager.gtest.xml'
    -- run_test.py: verify result file '/home/runner/work/ros2_control/ros2_control/.work/target_ws/build/controller_manager/test_results/controller_manager/test_controller_manager.gtest.xml'
  >>>
build/controller_manager/test_results/controller_manager/test_controller_manager.gtest.xml: 47 tests, 0 errors, 1 failure, 0 skipped
- controller_manager.per_controller_update_rate_with_different_cm_rates/TestControllerManagerWithDifferentCmUpdateRates per_controller_update_rate_with_different_cm_rates/4
  <<< failure message
    /home/runner/work/ros2_control/ros2_control/.work/target_ws/src/ros2_control/controller_manager/test/test_controller_manager.cpp:3718
    The difference between actual_updates and ctrl_update_rate is 12, which exceeds 1u, where
    actual_updates evaluates to 112,
    ctrl_update_rate evaluates to 100, and
    1u evaluates to 1.
    After 1000 CM cycles at 1000 Hz, a controller at 100 Hz should have updated exactly 100 times
  >>>

```
